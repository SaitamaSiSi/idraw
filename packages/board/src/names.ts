const _canvas = Symbol('_canvas');
const _displayCanvas = Symbol('_displayCanvas');
const _helperCanvas = Symbol('_helperCanvas');
const _mount = Symbol('_mount');
const _opts = Symbol('_opts');
const _hasRendered = Symbol('_hasRendered');
const _ctx = Symbol('_ctx');
const _helperCtx = Symbol('_helperCtx');

const _watcher = Symbol('_watcher');
const _render = Symbol('_render');
const _parsePrivateOptions = Symbol('_parsePrivateOptions');
const _scroller = Symbol('_scroller');
const _initEvent = Symbol('_initEvent');
const _doScrollX = Symbol('_doScrollX');
const _doScrollY = Symbol('_doScrollY');
const _doMoveScroll = Symbol('_doMoveScroll');
// const _doDoubleClick = Symbol('_doDoubleClick');
const _resetContext = Symbol('_resetContext');
const _screen = Symbol('_screen');
const _tempData = Symbol('_tempData');

export {
  _canvas, _displayCanvas, _mount, _opts, _hasRendered, _ctx,
  _watcher, _render, _parsePrivateOptions, _scroller,
  _initEvent, _doScrollX, _doScrollY, _doMoveScroll, _resetContext,
  _screen, _tempData, _helperCtx,  _helperCanvas,
};