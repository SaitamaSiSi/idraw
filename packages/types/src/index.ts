export * from './lib/data';
export * from './lib/context';
export * from './lib/board';
export * from './lib/paint';
export * from './lib/element';
export * from './lib/helper';
export * from './lib/config';
export * from './lib/core';
export * from './lib/screen';
export * from './lib/device';
export * from './lib/plugin';
export * from './lib/common';