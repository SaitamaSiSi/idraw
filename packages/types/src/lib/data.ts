import { DataElemDesc, DataElement, DataElementBase } from './element';

type IDrawDataBase = {
  elements: DataElementBase<keyof DataElemDesc>[];
  bgColor?: string;
};

type IDrawData = {
  elements: DataElement<keyof DataElemDesc>[];
  bgColor?: string;
};

export { IDrawData, IDrawDataBase };
