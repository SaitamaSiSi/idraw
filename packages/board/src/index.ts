import { 
  TypeScreenPosition, TypeScreenSize, TypeScreenContext, TypePoint, TypePointCursor,
  TypeBoardOptions, TypeBoardSizeOptions, TypeContext, 
} from 'idraw_zyh_types';
// tempTest
// import { 
//   TypeScreenPosition, TypeScreenSize, TypeScreenContext, TypePoint, TypePointCursor,
//   TypeBoardOptions, TypeBoardSizeOptions, TypeContext, 
// } from '../../types/src/index';
import util from 'idraw_zyh_util';
import { ScreenWatcher } from './lib/screen-watcher';
import { setStyle } from './lib/style';
import { TypeBoardEventArgMap } from './lib/event';
import { Scroller } from './lib/scroller';
import { Screen } from './lib/screen';
// import { TempData } from './lib/temp';

const { throttle, Context } = util;

import {
  _canvas, _displayCanvas, _mount, _opts, _hasRendered, _ctx,
  _watcher, _render, _parsePrivateOptions, _scroller, _helperCanvas, _helperCtx,
  _initEvent, _doScrollX, _doScrollY, _doMoveScroll, _resetContext,
  _screen, _tempData
} from './names';

type PrivateOptions = TypeBoardOptions & {
  devicePixelRatio: number
}

export class Board {

  private [_hasRendered] = false;

  private [_canvas]: HTMLCanvasElement;
  private [_helperCanvas]: HTMLCanvasElement;
  private [_displayCanvas]: HTMLCanvasElement;
  private [_mount]: HTMLDivElement;
  private [_opts]: PrivateOptions;
  private [_ctx]: TypeContext;
  private [_helperCtx]: TypeContext;
  // private [_watcher]: Watcher;
  private [_watcher]: ScreenWatcher;
  private [_scroller]: Scroller;
  private [_screen]: Screen;
  // private [_tempData]: TempData;

  constructor(mount: HTMLDivElement, opts: TypeBoardOptions) {
    // this[_tempData] = new TempData(opts);

    this[_mount] = mount;
    this[_canvas] = document.createElement('canvas');
    this[_helperCanvas] = document.createElement('canvas');
    this[_displayCanvas] = document.createElement('canvas');
    this[_mount].appendChild(this[_displayCanvas]);
    this[_opts] = this[_parsePrivateOptions](opts);

    const originCtx2d = this[_canvas].getContext('2d') as CanvasRenderingContext2D;
    const displayCtx2d = this[_displayCanvas].getContext('2d') as CanvasRenderingContext2D;
    const helperCtx2d = this[_helperCanvas].getContext('2d') as CanvasRenderingContext2D;
    this[_ctx] = new Context(originCtx2d, this[_opts]);
    this[_helperCtx] = new Context(helperCtx2d, this[_opts]);
    this[_screen] = new Screen(this[_ctx], this[_opts]);
    // this[_watcher] = new Watcher(this[_displayCanvas]);
    this[_watcher] = new ScreenWatcher(this[_displayCanvas], this[_ctx]);
    this[_scroller] = new Scroller(
      displayCtx2d, {
        width: opts.width,
        height: opts.height,
        devicePixelRatio: opts.devicePixelRatio || 1,
        scrollConfig: opts.scrollConfig,
      });
    this[_render]();
  }

  getDisplayContext2D(): CanvasRenderingContext2D {
    return this[_displayCanvas].getContext('2d') as CanvasRenderingContext2D;
  }

  getOriginContext2D(): CanvasRenderingContext2D {
    return this[_ctx].getContext();
  }

  getHelperContext2D(): CanvasRenderingContext2D {
    return this[_helperCtx].getContext();
  }

  getContext(): TypeContext {
    return this[_ctx];
  }

  getHelperContext(): TypeContext {
    return this[_helperCtx];
  }

  scale(scaleRatio: number): TypeScreenContext {
    if (scaleRatio > 0) {
      this[_ctx].setTransform({ scale: scaleRatio });
      this[_helperCtx].setTransform({ scale: scaleRatio });
    }
    const { position, size } = this[_screen].calcScreen();
    return { position, size};
  }

  scrollX(x: number) {
    this[_watcher].setStatusMap({
      canScrollYPrev: true,
      canScrollYNext: true,
      canScrollXPrev: true,
      canScrollXNext: true,
    })
    if (x >= 0 || x < 0) {
      this[_ctx].setTransform({ scrollX: x });
      this[_helperCtx].setTransform({ scrollX: x })
    }
    const { position, size, canScrollXNext, canScrollYNext, canScrollXPrev, canScrollYPrev } = this[_screen].calcScreen();
    this[_watcher].setStatusMap({
      canScrollYPrev,
      canScrollYNext,
      canScrollXPrev,
      canScrollXNext,
    })
    return { position, size};
  }

  scrollY(y: number): TypeScreenContext {
    this[_watcher].setStatusMap({
      canScrollYPrev: true,
      canScrollYNext: true,
      canScrollXPrev: true,
      canScrollXNext: true,
    })
    if (y >= 0 || y < 0) {
      this[_ctx].setTransform({ scrollY: y });
      this[_helperCtx].setTransform({ scrollY: y });
    }
    const { position, size, canScrollXNext, canScrollYNext, canScrollXPrev, canScrollYPrev } = this[_screen].calcScreen();
    this[_watcher].setStatusMap({
      canScrollYPrev,
      canScrollYNext,
      canScrollXPrev,
      canScrollXNext,
    })
    return { position, size};
  }

  getTransform() {
    return this[_ctx].getTransform();
  }

  draw(): TypeScreenContext {
    this.clear();
    const { position, deviceSize, size } = this[_screen].calcScreen();
    const displayCtx = this[_displayCanvas].getContext('2d');
    displayCtx?.drawImage(
      this[_canvas], deviceSize.x, deviceSize.y, deviceSize.w, deviceSize.h
    );
    displayCtx?.drawImage(
      this[_helperCanvas], deviceSize.x, deviceSize.y, deviceSize.w, deviceSize.h
    );
    if (this[_opts].canScroll === true) {
      this[_scroller].draw(position);
    }
    return { position, size };
  }

  clear() {
    const displayCtx = this[_displayCanvas].getContext('2d');
    displayCtx?.clearRect(0, 0, this[_displayCanvas].width, this[_displayCanvas].height);
  }

  on<T extends keyof TypeBoardEventArgMap >(name: T, callback: (p: TypeBoardEventArgMap[T]) => void) {
    this[_watcher].on(name, callback);
  }

  off<T extends keyof TypeBoardEventArgMap >(name: T, callback: (p: TypeBoardEventArgMap[T]) => void) {
    this[_watcher].off(name, callback);
  }

  getScreenInfo(): {
    size: TypeScreenSize, position: TypeScreenPosition, deviceSize: TypeScreenSize,
    width: number, height: number, devicePixelRatio: number
    } {
    return this[_screen].calcScreen();
  }

  setCursor(cursor: TypePointCursor) {
    this[_displayCanvas].style.cursor = cursor;
  }

  resetCursor() {
    this[_displayCanvas].style.cursor = 'auto';
  }

  resetSize(opts: TypeBoardSizeOptions) {
    this[_opts] = { ...this[_opts], ...opts };
    this[_resetContext]();
    this[_ctx].resetSize(opts);
    this[_helperCtx].resetSize(opts);
    this[_screen].resetSize(opts);
    this[_scroller].resetSize({
      width: this[_opts].width,
      height: this[_opts].height,
      devicePixelRatio: this[_opts].devicePixelRatio
    });
    this.draw();
  }

  getScrollLineWidth(): number {
    let lineWidth = 0;
    if (this[_opts].canScroll === true) {
      lineWidth = this[_scroller].getLineWidth();
    }
    return lineWidth;
  }

  pointScreenToContext(screenPoint: TypePoint): TypePoint {
    const { scrollX, scrollY, scale } = this.getTransform();
    const ctxPoint = {
      x: (screenPoint.x - scrollX) / scale,
      y: (screenPoint.y - scrollY) / scale,
    };
    return ctxPoint;
  }

  pointContextToScreen(ctxPoint: TypePoint): TypePoint {
    const { scrollX, scrollY, scale } = this.getTransform();
    const screenPoint = {
      x: ctxPoint.x * scale + scrollX,
      y: ctxPoint.y * scale + scrollY,
    };
    return screenPoint;
  }


  private [_render]() {
    if (this[_hasRendered] === true) {
      return;
    }
    this[_resetContext]();
    this[_initEvent]();
    this[_hasRendered] = true;
  }

  private [_resetContext] () {
    const { width, height, contextWidth, contextHeight, devicePixelRatio } = this[_opts];
    this[_canvas].width = contextWidth * devicePixelRatio;
    this[_canvas].height = contextHeight * devicePixelRatio;

    this[_helperCanvas].width = contextWidth * devicePixelRatio;
    this[_helperCanvas].height = contextHeight * devicePixelRatio;

    this[_displayCanvas].width = width * devicePixelRatio;
    this[_displayCanvas].height = height * devicePixelRatio;

    setStyle(this[_displayCanvas], {
      width: `${width}px`,
      height: `${height}px`,
    });
  }
  
  private [_parsePrivateOptions](opts: TypeBoardOptions): PrivateOptions {
    const defaultOpts = {
      devicePixelRatio: 1,
    };
    return { ...defaultOpts, ...opts };
  }

  private [_initEvent]() {
    if (this[_hasRendered] === true) {
      return;
    }
    if (this[_opts].canScroll === true) {

      this.on('wheelX', throttle((deltaX: number) => {
        this[_doScrollX](deltaX);
      }, 16));
      this.on('wheelY', throttle((deltaY: number) => {
        this[_doScrollY](deltaY);
      }, 16));

      let scrollType: 'x' | 'y' | null = null;
      this.on('moveStart', throttle((p: TypePoint) => {
        if (this[_scroller].isPointAtScrollX(p)) {
          scrollType = 'x';
        } else if (this[_scroller].isPointAtScrollY(p)) {
          scrollType = 'y';
        }
      }, 16));

      this.on('move', throttle((p: TypePoint) => {
        if (scrollType) {
          this[_doMoveScroll](scrollType, p);
        }
      }, 16));

      this.on('moveEnd', throttle((p: TypePoint) => {
        if (scrollType) {
          this[_doMoveScroll](scrollType, p);
        }
        scrollType = null;
      }, 16));

      // this.on('doubleClick', (p: TypePoint) => {})
    }
  }

  private [_doScrollX](dx: number, prevScrollX?: number) {
    const { width } = this[_opts];
    let scrollX = prevScrollX;
    if (!(typeof scrollX === 'number' && (scrollX > 0 || scrollX <= 0))) {
      scrollX = this[_ctx].getTransform().scrollX;
    }
    const { position } = this[_screen].calcScreen();
    const { xSize } = this[_scroller].calc(position);
    const moveX = this[_screen].calcScreenScroll(position.left, position.right, xSize, width, dx);
    this.scrollX(scrollX + moveX);
    this.draw();
  }

  private [_doScrollY](dy: number, prevScrollY?: number) {
    const { height } = this[_opts];
    let scrollY = prevScrollY;
    if (!(typeof scrollY === 'number' && (scrollY > 0 || scrollY <= 0))) {
      scrollY = this[_ctx].getTransform().scrollY;
    }
    const { position } = this[_screen].calcScreen();
    const { ySize } = this[_scroller].calc(position);
    const moveY = this[_screen].calcScreenScroll(position.top, position.bottom, ySize, height, dy);
    this.scrollY(scrollY + moveY);
    this.draw();
  }

  private [_doMoveScroll](scrollType: 'x' | 'y', point: TypePoint) {
    if (!scrollType) {
      return;
    }
    const { position } = this[_screen].calcScreen();
    const { xSize, ySize } = this[_scroller].calc(position);
    if (scrollType === 'x') {
      this[_doScrollX](point.x - xSize / 2, 0);
    } else if (scrollType === 'y') {
      this[_doScrollY](point.y - ySize / 2, 0);
    }
  }


}

 

