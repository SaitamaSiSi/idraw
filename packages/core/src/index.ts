import {
  IDrawData,
  Point,
  BoardSizeOptions,
  IDrawConfig,
  IDrawConfigStrict,
  DataElementBase,
  DataElement,
  DataElemDesc,
  IDrawContext,
  CoreOptions,
  ScreenContext,
  ScreenData
} from '@idraw/types';
import Board from '@idraw/board';
import { deepClone } from '@idraw/util';
import Renderer from '@idraw/renderer';
import is, { IsTypeUtil } from './lib/is';
import check, { CheckTypeUtil } from './lib/check';
import {
  Element,
  mergeConfig,
  CoreEvent,
  TypeCoreEventArgMap,
  parseData,
  TempData,
  diffElementResourceChangeList
} from './lib';
import {
  getSelectedElements,
  updateElement,
  selectElementByIndex,
  selectElement,
  cancelElementByIndex,
  cancelElement,
  getElement,
  getElementByIndex,
  moveUpElement,
  moveDownElement,
  addElement,
  deleteElement,
  insertElementBefore,
  insertElementBeforeIndex,
  insertElementAfter,
  insertElementAfterIndex
} from './mixins/element';
// import { initEvent } from './mixins/event';
import { Engine } from './lib/engine';
import {
  drawElementWrapper,
  drawAreaWrapper,
  drawElementListWrappers
} from './lib/draw/wrapper';

export default class Core {
  $data: IDrawData;
  private _board: Board;
  private _opts: CoreOptions;
  private _config: IDrawConfigStrict;
  private _renderer: Renderer;
  private _elementHandler: Element;
  private _coreEvent: CoreEvent = new CoreEvent();
  private _tempData: TempData = new TempData();
  private _engine: Engine;

  static is: IsTypeUtil = is;
  static check: CheckTypeUtil = check;

  constructor(mount: HTMLDivElement, opts: CoreOptions, config?: IDrawConfig) {
    this.$data = { elements: [] };
    this._opts = opts;
    this._config = mergeConfig(config || {});
    this._board = new Board(mount, {
      ...this._opts,
      canScroll: config?.scrollWrapper?.use,
      scrollConfig: {
        color: config?.scrollWrapper?.color || '#000000',
        width: config?.scrollWrapper?.width || 12,
        ...(config?.scrollWrapper || {})
      }
    });
    this._renderer = new Renderer();
    const drawFrame = () => {
      const helperCtx = this._board.getHelperContext();
      const helperConfig = this._engine.getHelperConfig();
      this._board.clear();
      const { contextWidth, contextHeight, devicePixelRatio } = this._opts;
      helperCtx.clearRect(
        0,
        0,
        contextWidth * devicePixelRatio,
        contextHeight * devicePixelRatio
      );
      drawElementWrapper(helperCtx, helperConfig);
      drawAreaWrapper(helperCtx, helperConfig);
      drawElementListWrappers(helperCtx, helperConfig);
      this._board.draw();
    };
    this._renderer.on('drawFrame', () => {
      drawFrame();
    });
    this._renderer.on('drawFrameComplete', () => {
      drawFrame();
    });
    this._elementHandler = new Element(this._board.getContext());
    this._engine = new Engine({
      coreEvent: this._coreEvent,
      board: this._board,
      element: this._elementHandler,
      config: this._config,
      drawFeekback: this.$draw.bind(this),
      getDataFeekback: () => this.$data,
      selectElementByIndex: this.selectElementByIndex.bind(this),
      emitChangeScreen: this._emitChangeScreen.bind(this),
      emitChangeData: this.$emitChangeData.bind(this)
    });
    this._engine.init();

    this._renderer.on('drawFrame', () => {
      this._coreEvent.trigger('drawFrame', undefined);
    });
    this._renderer.on('drawFrameComplete', () => {
      this._coreEvent.trigger('drawFrameComplete', undefined);
    });

    this._tempData.set('hasInited', true);
  }

  private _emitChangeScreen() {
    if (this._coreEvent.has('changeScreen')) {
      this._coreEvent.trigger('changeScreen', {
        ...this.getScreenTransform()
      });
    }
  }

  $draw(opts?: { resourceChangeUUIDs?: string[] }): void {
    this._engine.updateHelperConfig({
      width: this._opts.width,
      height: this._opts.height,
      devicePixelRatio: this._opts.devicePixelRatio
    });

    this._renderer.thaw();
    this._renderer.render(this._board.getContext(), this.$data, {
      changeResourceUUIDs: opts?.resourceChangeUUIDs || []
    });
  }

  getElement(uuid: string) {
    return getElement(this, uuid);
  }

  getElementByIndex(index: number) {
    return getElementByIndex(this, index);
  }

  selectElementByIndex(index: number): void {
    return selectElementByIndex(this, index);
  }

  selectElement(uuid: string): void {
    return selectElement(this, uuid);
  }

  cancelElementByIndex(index: number): void {
    return cancelElementByIndex(this, index);
  }

  cancelElement(uuid: string): void {
    return cancelElement(this, uuid);
  }

  moveUpElement(uuid: string): void {
    return moveUpElement(this, uuid);
  }

  moveDownElement(uuid: string): void {
    return moveDownElement(this, uuid);
  }

  updateElement(elem: DataElement<keyof DataElemDesc>) {
    return updateElement(this, elem);
  }

  addElement(elem: DataElementBase<keyof DataElemDesc>): string | null {
    return addElement(this, elem);
  }

  deleteElement(uuid: string) {
    return deleteElement(this, uuid);
  }

  insertElementBefore(
    elem: DataElementBase<keyof DataElemDesc>,
    beforeUUID: string
  ) {
    return insertElementBefore(this, elem, beforeUUID);
  }

  insertElementBeforeIndex(
    elem: DataElementBase<keyof DataElemDesc>,
    index: number
  ) {
    return insertElementBeforeIndex(this, elem, index);
  }

  getSelectedElements() {
    return getSelectedElements(this);
  }

  insertElementAfter(
    elem: DataElementBase<keyof DataElemDesc>,
    beforeUUID: string
  ) {
    return insertElementAfter(this, elem, beforeUUID);
  }

  insertElementAfterIndex(
    elem: DataElementBase<keyof DataElemDesc>,
    index: number
  ) {
    return insertElementAfterIndex(this, elem, index);
  }

  resetSize(opts: BoardSizeOptions) {
    this._opts = { ...this._opts, ...opts };
    this._board.resetSize(opts);
    this.$draw();
  }

  scale(ratio: number): ScreenContext {
    const screen = this._board.scale(ratio);
    this.$draw();
    this._emitChangeScreen();
    return screen;
  }

  scrollLeft(left: number): ScreenContext {
    const screen = this._board.scrollX(0 - left);
    this.$draw();
    this._emitChangeScreen();
    return screen;
  }

  scrollTop(top: number): ScreenContext {
    const screen = this._board.scrollY(0 - top);
    this.$draw();
    this._emitChangeScreen();
    return screen;
  }

  getScreenTransform(): ScreenData {
    const transform = this._board.getTransform();
    return {
      scale: transform.scale,
      scrollTop: Math.max(0, 0 - transform.scrollY),
      scrollLeft: Math.max(0, 0 - transform.scrollX)
    };
  }

  getData(): IDrawData {
    return deepClone(this.$data);
  }

  setData(data: any | IDrawData, opts?: { triggerChangeEvent: boolean }): void {
    const resourceChangeUUIDs = diffElementResourceChangeList(this.$data, data);
    this.$data = this._elementHandler.initData(deepClone(parseData(data)));
    if (opts && opts.triggerChangeEvent === true) {
      this.$emitChangeData();
    }
    this.$draw({ resourceChangeUUIDs });
  }

  clearOperation() {
    this._tempData.clear();
    this.$draw();
  }

  on<T extends keyof TypeCoreEventArgMap>(
    key: T,
    callback: (p: TypeCoreEventArgMap[T]) => void
  ) {
    this._coreEvent.on(key, callback);
  }

  off<T extends keyof TypeCoreEventArgMap>(
    key: T,
    callback: (p: TypeCoreEventArgMap[T]) => void
  ) {
    this._coreEvent.off(key, callback);
  }

  getEngine() {
    return this._engine;
  }

  pointScreenToContext(p: Point) {
    return this._board.pointScreenToContext(p);
  }

  pointContextToScreen(p: Point) {
    return this._board.pointContextToScreen(p);
  }

  $getBoardContext(): IDrawContext {
    return this._board.getContext();
  }

  $getDisplayContext2D(): CanvasRenderingContext2D {
    return this._board.getDisplayContext2D();
  }

  $getOriginContext2D(): CanvasRenderingContext2D {
    return this._board.getOriginContext2D();
  }

  $emitChangeData() {
    if (this._coreEvent.has('changeData')) {
      this._coreEvent.trigger('changeData', deepClone(this.$data));
    }
  }

  $getElementHandler() {
    return this._elementHandler;
  }
}
