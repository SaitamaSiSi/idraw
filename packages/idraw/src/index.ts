import Core from '@idraw/core';
import { IDrawData, IDrawConfig } from '@idraw/types';
import { Options, PrivateOptions } from './types';
import { defaultOptions } from './config';
import { TempData } from './lib/temp';
import { KeyboardWatcher } from './lib/keyboard-watcher';

import { redo, undo } from './mixins/record';
import { exportDataURL, toDataURL } from './mixins/file';
import {
  copyElements,
  pasteElements,
  cutElements,
  deleteElements,
  keyArrowUp,
  keyArrowDown,
  keyArrowLeft,
  keyArrowRight,
  keyUndo
} from './mixins/keyboard';
// import { version } from './../package.json';

export default class iDraw extends Core {
  private _opts: PrivateOptions;
  private _hasInited = false;
  private _tempData = new TempData();
  private _keyboardWatcher = new KeyboardWatcher();

  // static version = version;

  constructor(mount: HTMLDivElement, opts: Options, config?: IDrawConfig) {
    super(
      mount,
      {
        width: opts.width || defaultOptions.width,
        height: opts.height || defaultOptions.height,
        contextWidth: opts.contextWidth || defaultOptions.contextWidth,
        contextHeight: opts.contextHeight || defaultOptions.contextHeight,
        devicePixelRatio:
          opts.devicePixelRatio || defaultOptions.devicePixelRatio,
        onlyRender: opts.onlyRender || defaultOptions.onlyRender
      },
      config || {}
    );
    this._opts = this._createOpts(opts);
    this._initEvent();
  }

  undo(): { doRecordCount: number; data: IDrawData | null } {
    return undo(this);
  }

  redo(): { undoRecordCount: number; data: IDrawData | null } {
    return redo(this);
  }

  toDataURL(type: 'image/png' | 'image/jpeg', quality?: number): string {
    return toDataURL(this, type, quality);
  }

  getTempData() {
    return this._tempData;
  }

  async exportDataURL(
    type: 'image/png' | 'image/jpeg',
    quality?: number
  ): Promise<string> {
    return exportDataURL(this, type, quality);
  }

  private _initEvent() {
    if (this._hasInited === true) {
      return;
    }
    this.on('changeData', (data: IDrawData) => {
      this._pushRecord(data);
    });
    this.on('mouseLeaveScreen', () => {
      this._tempData.set('isFocus', false);
    });
    this.on('mouseOverScreen', () => {
      this._tempData.set('isFocus', true);
    });
    if (this._opts.disableKeyboard === false) {
      this._keyboardWatcher
        .on('keyboardCopy', () => copyElements(this))
        .on('keyboardPaste', () => pasteElements(this))
        .on('keyboardCut', () => cutElements(this))
        .on('keyboardDelete', () => deleteElements(this))
        .on('keyboardArrowUp', () => keyArrowUp(this))
        .on('keyboardArrowDown', () => keyArrowDown(this))
        .on('keyboardArrowLeft', () => keyArrowLeft(this))
        .on('keyboardArrowRight', () => keyArrowRight(this))
        .on('keyboardUndo', () => keyUndo(this));
    }
    this._hasInited = true;
  }

  private _pushRecord(data: IDrawData) {
    const doRecords = this._tempData.get('doRecords');
    if (doRecords.length >= this._opts.maxRecords) {
      doRecords.shift();
    }
    doRecords.push({ data, time: Date.now() });
    this._tempData.set('doRecords', doRecords);
    this._tempData.set('unDoRecords', []);
  }

  private _createOpts(opts: Options): PrivateOptions {
    return { ...{}, ...defaultOptions, ...opts };
  }
}
