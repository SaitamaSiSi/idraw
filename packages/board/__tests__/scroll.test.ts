import { Board } from '../src';
import { getData } from './data';

describe('@idraw/board', () => {


  document.body.innerHTML = `
    <div id="mount"></div>
  `;
  const opts = {
    width: 600,
    height: 400,
    contextWidth: 600,
    contextHeight: 400,
    devicePixelRatio: 4
  }
  const mount = document.querySelector('#mount') as HTMLDivElement;
  const board = new Board(mount, opts);

  const ctx = board.getContext();
  const data = getData();
  board.clear();
  ctx.clearRect(0, 0, opts.width, opts.height);
  ctx.setFillStyle('#ffffff');
  ctx.fillRect(0, 0, opts.width, opts.height);
  data.elements.forEach(ele => {
    ctx.setFillStyle(ele.desc.color);
    ctx.fillRect(ele.x, ele.y, ele.w, ele.h);
  });

  test('scroll', async () => {  
    
  
    const resultScale =board.scale(2);
    expect(resultScale).toStrictEqual({"position":{"top":0,"bottom":-400,"left":0,"right":-600},"size":{"x":0,"y":0,"w":1200,"h":800}})
  
    const resultX =board.scrollX(-600);
    expect(resultX).toStrictEqual({"position":{"top":0,"bottom":-400,"left":-600,"right":0},"size":{"x":-1200,"y":0,"w":1200,"h":800}})
  
    const resultY =board.scrollY(-400); 
    expect(resultY).toStrictEqual({"position":{"top":-400,"bottom":0,"left":-600,"right":0},"size":{"x":-1200,"y":-800,"w":1200,"h":800}})
  
    board.draw();
  
    const originCtx = board.getOriginContext2D();
    // @ts-ignore;
    const originCalls = originCtx.__getDrawCalls();
    expect(originCalls).toMatchSnapshot();
  
    const displayCtx = board.getDisplayContext2D();
    // @ts-ignore;
    const displayCalls = displayCtx.__getDrawCalls();
    expect(displayCalls).toMatchSnapshot();


    const screenInfo = board.getScreenInfo();
    expect(screenInfo).toStrictEqual({"size":{"x":-1200,"y":-800,"w":1200,"h":800},"position":{"top":-400,"bottom":0,"left":-600,"right":0},"deviceSize":{"x":-2400,"y":-1600,"w":4800,"h":3200},"width":600,"height":400,"devicePixelRatio":4, "canScrollXNext": true, "canScrollXPrev": true, "canScrollYNext": true, "canScrollYPrev": true, });

  });


});

