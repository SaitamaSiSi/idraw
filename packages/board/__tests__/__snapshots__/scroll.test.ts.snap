// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`@idraw/board scroll 1`] = `
[
  {
    "props": {
      "height": 1600,
      "width": 2400,
      "x": 0,
      "y": 0,
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "clearRect",
  },
  {
    "props": {
      "height": 1600,
      "width": 2400,
      "x": 0,
      "y": 0,
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillRect",
  },
  {
    "props": {
      "height": 480,
      "width": 800,
      "x": 40,
      "y": 40,
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillRect",
  },
  {
    "props": {
      "height": 480,
      "width": 800,
      "x": 320,
      "y": 320,
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillRect",
  },
  {
    "props": {
      "height": 480,
      "width": 800,
      "x": 640,
      "y": 640,
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillRect",
  },
  {
    "props": {
      "height": 400,
      "width": 800,
      "x": 1560,
      "y": 1160,
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillRect",
  },
]
`;

exports[`@idraw/board scroll 2`] = `
[
  {
    "props": {
      "height": 1600,
      "width": 2400,
      "x": 0,
      "y": 0,
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "clearRect",
  },
  {
    "props": {
      "height": 1600,
      "width": 2400,
      "x": 0,
      "y": 0,
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "clearRect",
  },
  {
    "props": {
      "dHeight": 1600,
      "dWidth": 2400,
      "dx": -2400,
      "dy": -1600,
      "img": <canvas
        height="1600"
        width="2400"
      />,
      "sHeight": 1600,
      "sWidth": 2400,
      "sx": 0,
      "sy": 0,
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "drawImage",
  },
  {
    "props": {
      "dHeight": 1600,
      "dWidth": 2400,
      "dx": -2400,
      "dy": -1600,
      "img": <canvas
        height="1600"
        width="2400"
      />,
      "sHeight": 1600,
      "sWidth": 2400,
      "sx": 0,
      "sy": 0,
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "drawImage",
  },
]
`;
